### PR TITLE
Split signup providers into public vs invite-gated lists

### PR DESCRIPTION
**Pull request by brendanlong**
## Problem

Public signups are disabled on the production instance, but even **with** a valid invite you can only sign up via Google/Apple — there's no way to create an email/password account (e.g. for a bot).

The cause: two orthogonal flags governed signups and didn't compose:

- `ALLOW_ALL_SIGNUPS` — is an invite required?
- `ALLOWED_SIGNUP_PROVIDERS` — which providers are allowed, **regardless** of invite.

The provider check ran before the invite logic (`createUser` in `signup.ts`), so an invite could never unlock a provider the allowlist excluded. "Email allowed, but only with an invite" was impossible to express.

## Change

Replace `ALLOW_ALL_SIGNUPS` with a second provider list, so each provider is independently **public**, **invite-only**, or **denied**:

| Env var | Default | Meaning |
| --- | --- | --- |
| `ALLOWED_SIGNUP_PROVIDERS` | all | Providers allowed to sign up **with** a valid invite |
| `ALLOWED_PUBLIC_SIGNUP_PROVIDERS` | none | Subset allowed **without** an invite |

- Empty public list = fully invite-only (the old `ALLOW_ALL_SIGNUPS=false`).
- Public list = every provider = the old `ALLOW_ALL_SIGNUPS=true`.
- The public list is intersected with the allowlist, so it can never widen access beyond it.

A new pure helper `resolveSignupProviderAccess(provider, lists)` returns `public | invite-only | denied` and is shared by the signup service (enforcement) and the `signupConfig` endpoint (UI description). The register page now chooses the effective provider list based on whether an invite is present in the URL; `signupConfig` additionally returns `publicSignupProviders`, and `requiresInvite` now means "no provider is public".

### Production behavior (`fly.toml`)

```toml
ALLOWED_SIGNUP_PROVIDERS = "apple,google,email"   # email allowed with an invite
ALLOWED_PUBLIC_SIGNUP_PROVIDERS = "apple,google"  # public OAuth only, no public email
```

This **preserves** current public behavior (apple/google open, no public email signup) and **enables** invite-gated email/bot accounts. Deploying picks up the new env automatically; the old `ALLOW_ALL_SIGNUPS` secret can be removed.

## Testing

- `pnpm typecheck`, `pnpm lint` — clean
- `pnpm test:unit` — 1890 pass, including new `tests/unit/signup-providers.test.ts` covering the resolver (public/invite-only/denied, empty + full public lists)
- Default behavior is provably unchanged (no invite → invite-only), and no existing test sets these env vars, so integration/e2e behavior is unaffected. Integration/e2e were not run locally (no Docker in this environment).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/server/config/env.ts [hunk 1]
```diff
@@ -13,44 +13,74 @@ export const ALL_SIGNUP_PROVIDERS = ["email", "google", "apple", "discord"] as c
...
-  if (!raw) return ALL_SIGNUP_PROVIDERS;
-
-  const providers = raw
+function parseProviderList(raw: string | undefined): SignupProvider[] {
+  if (!raw) return [];
+  return raw
     .split(",")
     .map((s) => s.trim().toLowerCase())
     .filter((s): s is SignupProvider => ALL_SIGNUP_PROVIDERS.includes(s as SignupProvider));
+}
```
[the remark this answers is not in the document]
Comment (brendanlong) [reply src/server/config/env.ts:28]: Good catch — fixed in 39412aa2. `parseProviderList` now collapses duplicates via a `Set` before they can reach the client and cause duplicate-key warnings in `OAuthButtons`. Added unit tests covering dedupe plus the other parsing edge cases (trim/lowercase, unknown-drop, public∩allowed intersection, set-but-all-invalid fallback).

— Claude
